ray-curator/api-design.md

```diff
@@ -0,0 +1,196 @@
+
+
+## Core Components
+
+1. **Tasks** : The fundamental unit of data that flows through the curation pipeline.
+2. **Stages** : The fundamental processing unit that takes Tasks as input and produces Tasks as output.
+3. **Pipelines** : A collection of stages that defines the complete processing workflow.
+4. **Executors** : The component that orchestrates the execution of the pipeline.
+
+### Tasks
+
+A **Task** is the fundamental unit of data that flows through the curation pipeline, representing a batch of input data for processing.
+
+#### Base Task Implementation
+
+```python
+@dataclass
+class Task(ABC, Generic[T]):
+    """Abstract base class for tasks in the pipeline."""
+    task_id: str
+    dataset_name: str
+    data: T
+    _stage_perf: list[StagePerfStats] = field(default_factory=list)
+    _metadata: dict[str, Any] = field(default_factory=dict)
+
+    @property
+    @abstractmethod
+    def num_items(self) -> int:
+        """Get the number of items in this task."""
+
+    @abstractmethod
+    def validate(self) -> bool:
+        """Validate the task data."""
+```
+
+#### Example Task Types
+
+#### DocumentTask
+For text-based curation pipelines:
+
+```python
+@dataclass
+class DocumentTask(Task[pd.DataFrame]):
+    """Task for document processing."""
+    
+    @property
+    def num_items(self) -> int:
+        return len(self.data)
+    
+    def validate(self) -> bool:
+        return isinstance(self.data, pd.DataFrame) and not self.data.empty
+```
+
+
+### Stages
+
+
+#### Base Stage Interface
+
+```python
+class ProcessingStage(ABC, Generic[X, Y], metaclass=StageMeta):
+    """Base class for all processing stages."""
+    
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """Unique name for this stage."""
+
+    @property
+    def resources(self) -> Resources:
+        """Resource requirements for this stage."""
+        return Resources(cpus=1.0)
+
+    @abstractmethod
+    def process(self, task: X) -> Y | list[Y]:
+        """Process a single task."""
+```
+
+#### Resource Specification
+
+```python
+@dataclass
+class Resources:
+    """Define resource requirements for a processing stage."""
+    cpus: float = 1.0 # Number of CPU cores
+    gpu_memory_gb: float = 0.0 # Number of GPU memory in GB (Only for single GPU)
+    gpus: float = 0.0 # Number of GPUs (Only for multi-GPU)
+    nvdecs: int = 0 # Number of NVDEC decoders
+    nvencs: int = 0 # Number of NVENC encoders
+    entire_gpu: bool = False # Whether to use the entire GPU
+```
+
+### Pipelines
+
+A **Pipeline** is a collection of stages that defines the complete processing workflow.
+
+```python
+class Pipeline:
+    """A pipeline defines a sequence of processing stages."""
+    
+    def __init__(self, stages: list[ProcessingStage]):
+        self.stages = stages
+    
+    def add_stage(self, stage: ProcessingStage):
+        """Add a stage to the pipeline."""
+    
+    def run(self, executor: BaseExecutor | None = None) -> list[Task]:
```

Comment:
nit
```suggestion
    def run(self, executor: BaseExecutor | None = None) -> list[Task] | None:
```
Because in theory an executor might be configurable to not return anything to the driver since that results in serialization overhead. There is a xenna param for the same, can't remember the exact one